File: 2222-number-of-ways-to-select-buildings/2222-number-of-ways-to-select-buildings.cpp
class Solution {
private: 
	long long dp[100003][3][4];

    long long solve(string& s, int idx, int prev, int len, int n) {
        if(len == 3) return 1;
        if(idx == n) {
            return 0;
        }
        if(dp[idx][prev][len] != -1) return dp[idx][prev][len];

        long long res = solve(s, idx+1, prev, len, n);

        if(prev != (s[idx]-'0')) {
            res += solve(s, idx+1, s[idx]-'0', len+1, n);
        }
        return dp[idx][prev][len] = res;
    }

public:    
    
    long long numberOfWays(string s) {
        int n = s.size();
        memset(dp, -1, sizeof(dp));
        return solve(s, 0, 2, 0, n);
    }
};

File: 2222-number-of-ways-to-select-buildings/2222-number-of-ways-to-select-buildings.py
class Solution:
    def numberOfWays(self, s: str) -> int:
        
        
        z, o, zo, oz, total = 0, 0, 0, 0, 0
        for c in s:
            if c == '1':
                total += oz
                zo += z
                o += 1
            elif c == '0':
                total += zo
                oz += o
                z += 1
        return total
        
        """
        clusters = [1]
        res = 0
        
        for i in range(1, len(s)):
            if s[i] == s[i-1]:
                clusters[-1] += 1
            else:
                clusters.append(1)
                
        n = len(clusters)
        for i in range(n):
            for j in range(i+1, n, 2):
                for k in range(j+1, n, 2):
                    res += clusters[i]*clusters[j]*clusters[k]
        return res
        
        """